### DISCUSSION TITLE
Replace current SQLite (2.6.0) with new WASM-native version?

### DISCUSSION THREAD

**Discussion by michaelweinold** [pyodide/pyodide#4981] [General]
Related:

 - https://github.com/emscripten-forge/recipes/issues/1254

The current `sqlite3` version shipped with Pyodide is `v2.6.0`. \
However, since version `v3.4.0`, a WASM-native build consisting of 

> A precompiled bundle of sqlite3.wasm and its JavaScript APIs, ready for use in web applications.

[is available for download](https://www.sqlite.org/download.html).

Would it be possible to switch to this version? This might resolve some of the current SQLite/WASM file-system compatibility issues documented here:

 - https://github.com/jupyterlite/pyodide-kernel/issues/35

**Comment by hoodmane** [comment]
> A precompiled bundle of sqlite3.wasm and its JavaScript APIs

This won't help us because we need to link it into an Emscripten shared libraray (or the main binary). They have their own loader code.

It sounds like the problem is that your file system is not compliant. Have you tried running the unistd test suite against your file system and fixing the test failures? I doubt updating sqlite will fix the problem since version 3.x will probably have almost the same expectations for the file system as version 2.x. Though if we did somehow use `sqlite3.wasm` it has its own entirely separate MEMFS which is more compliant.

**Reply by michaelweinold** [reply]
...thank you for letting me know!

From my understanding, [JupyterLite uses](https://jupyterlite.readthedocs.io/en/latest/howto/configure/storage.html#local-storage-drivers) IndexedDB [through localForage](https://github.com/localForage/localForage).

Unfortunately, testing the POSIX compliance of the different file-system components is beyond my abilities at the moment (I am likely even using incorrect terminology here - which tool/service/driver might be the root cause of the issue due to missing POSIX compliance? IndexedDB, the localForage driver, the Pyodide driveFS?) 

The current workaround reamains to store the `.db` file in the `/tmp` directory:

 - https://github.com/jupyterlite/pyodide-kernel/issues/35

**Reply by hoodmane** [reply]
I think the code that needs to be checked is the drivefs, which I guess is ultimately backed by localForage:
https://github.com/jupyterlite/jupyterlite/blob/main/docs/reference/contents.md
https://github.com/jupyterlite/jupyterlite/blob/main/packages/contents/src/

Unfortunately there seem to be many many layers of indirection here so there's a lot of different places that could be wrong.


Here's Emscripten's test suite for file systems:
https://github.com/emscripten-core/emscripten/tree/main/test/unistd

You would need some way to reset the state of the drive between tests.

**Comment by ryanking13** [comment]
I also think updating the sqlite3 version would not fix anything, but FYI, the libsqlite3 version used in pyodide v0.26 is [3.39.0 which comes from emscripten 3.1.58](https://github.com/emscripten-core/emscripten/blob/a41843e0860e52c948c1fce20307933c6631c800/tools/ports/sqlite3.py#L10)
